### PR TITLE
Refactor scalar IR nodes

### PR DESCRIPTION
**Pull request by naoyam**
Follow-up to #2197 and #2198

Removed `Float`.

`Double`, `Int` and `Bool` use `Scalar` template class.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_interface_nodes.h [hunk 1]
```diff
@@ -23,59 +23,59 @@ class ViewTransform;
...
+      : Val(passkey, ValType::Scalar, dtype), maybe_value_{value} {
+    TORCH_INTERNAL_ASSERT(
+        (std::is_integral<UnderlyingType>::value && isIntegralType(dtype)) ||
+            (std::is_same<UnderlyingType, bool>::value &&
+             isBooleanType(dtype)) ||
+            (std::is_floating_point<UnderlyingType>::value &&
+             isFloatingPointType(dtype)),
+        "Invalid data type: ",
+        dtype);
+  }
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_interface_nodes.h]: Could these two ctors be merge into a single one? I think only keeping the `c10::optional` version should be fine, and `UnderlyingType` could be automatically converted to `c10::optional<UnderlyingType>`

File: torch/csrc/jit/codegen/cuda/dispatch.h [hunk 2]
```diff
@@ -61,12 +61,11 @@ class IterDomain;
...
-template <DataType DT>
-class FloatingPoint;
-using Float = FloatingPoint<DataType::Float>;
-using Double = FloatingPoint<DataType::Double>;
-class Int;
+template <typename DT>
+class Scalar;
+using Bool = Scalar<bool>;
+using Double = Scalar<double>;
+using Int = Scalar<int64_t>;
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/dispatch.h:68]: This is a bit confusing. Should we change the name to `FloatingPoint` and `Integer`?

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 3]
```diff
@@ -241,12 +223,19 @@ void IrPrinter::handle(const Double* d) {
...
-        << std::setprecision(
-               std::numeric_limits<Double::ScalarType>::max_digits10)
-        << *(d->value()) << ")";
+    os_ << d->getDataType().value() << "(";
+    if (d->getDataType() == DataType::Double) {
+      os_ << std::setprecision(std::numeric_limits<double>::max_digits10)
+          << (float)*(d->value()) << ")";
+    } else if (d->getDataType() == DataType::Float) {
+      os_ << std::setprecision(std::numeric_limits<float>::max_digits10)
+          << *(d->value()) << ")";
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: Should this be the following?
```suggestion
    if (d->getDataType() == DataType::Double) {
      os_ << std::setprecision(std::numeric_limits<double>::max_digits10)
          << *(d->value()) << ")";
    } else if (d->getDataType() == DataType::Float) {
      os_ << std::setprecision(std::numeric_limits<float>::max_digits10)
          << (float)*(d->value()) << ")";
```

File: torch/csrc/jit/codegen/cuda/dispatch.h (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/dispatch.h:68]: I'm fine with the names, just didn't want to change the existing names. I'd prefer shorter nameas and keep `Int` as is and rename `Double` to `Float`. What do you think?

**Comment by zasdfgbnm** [comment]
The only blocking comment is https://github.com/csarofeen/pytorch/pull/2203#discussion_r1027008023, otherwise LGTM

File: torch/csrc/jit/codegen/cuda/dispatch.h (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/dispatch.h:68]: That sounds good to me

**Review by zasdfgbnm — APPROVED** [review]
Thanks for refactoring! I am wondering if I should start using `Scalar` to replace `Attribute` in exprs when applicable. It makes more sense to store a `Bool*` instead of a `Attribute<bool>*`

File: torch/csrc/jit/codegen/cuda/dispatch.h (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/dispatch.h:68]: Actually, `Float` seems a little off to me as the default underlying type is `double`, `Double` seems more intuitive. Let's keep the names as is.

**Comment by naoyam** [comment]
> Thanks for refactoring! I am wondering if I should start using `Scalar` to replace `Attribute` in exprs when applicable. It makes more sense to store a `Bool*` instead of a `Attribute<bool>*`

Yeah, I don't know if that would have any immediate benefit, but I guess it would make IR serialization a little easier.

**Pull request merged**